runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,773 @@
+  local matches = M._lsp_to_complete_items(result, prefix, client_id)
+  return matches, server_start_boundary
+end
+
+--- Implements 'omnifunc' compatible LSP completion.
+---
+---@see |complete-functions|
+---@see |complete-items|
+---@see |CompleteDone|
+---
+---@param findstart integer 0 or 1, decides behavior
+---@param base integer findstart=0, text to match against
+---
+---@return integer|table Decided by {findstart}:
+--- - findstart=0: column where the completion starts, or -2 or -3
+--- - findstart=1: list of matches (actually just calls |complete()|)
+function M.omnifunc(findstart, base)
+  vim.lsp.log.debug('omnifunc.findstart', { findstart = findstart, base = base })
+  assert(base) -- silence luals
+  local bufnr = api.nvim_get_current_buf()
+  local clients = lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_completion })
+  local remaining = #clients
+  if remaining == 0 then
+    return findstart == 1 and -1 or {}
+  end
+
+  local win = api.nvim_get_current_win()
+  local cursor = api.nvim_win_get_cursor(win)
+  local lnum = cursor[1] - 1
+  local cursor_col = cursor[2]
+  local line = api.nvim_get_current_line()
+  local line_to_cursor = line:sub(1, cursor_col)
+  local client_start_boundary = vim.fn.match(line_to_cursor, '\\k*$') --[[@as integer]]
+  local server_start_boundary = nil
+  local items = {}
+
+  local function on_done()
+    local mode = api.nvim_get_mode()['mode']
+    if mode == 'i' or mode == 'ic' then
+      vim.fn.complete((server_start_boundary or client_start_boundary) + 1, items)
+    end
+  end
+
+  local util = vim.lsp.util
+  for _, client in ipairs(clients) do
+    local params = util.make_position_params(win, client.offset_encoding)
+    client.request(ms.textDocument_completion, params, function(err, result)
+      if err then
+        lsp.log.warn(err.message)
+      end
+      if result and vim.fn.mode() == 'i' then
+        local matches
+        matches, server_start_boundary = M._convert_results(
+          line,
+          lnum,
+          cursor_col,
+          client.id,
+          client_start_boundary,
+          server_start_boundary,
+          result,
+          client.offset_encoding
+        )
+        vim.list_extend(items, matches)
+      end
+      remaining = remaining - 1
+      if remaining == 0 then
+        vim.schedule(on_done)
+      end
+    end, bufnr)
+  end
+
+  -- Return -2 to signal that we should continue completion so that we can
+  -- async complete.
+  return -2
+end
+
+--- Initializes the completion commands for the given client.
+---
+--- @param client vim.lsp.Client
+local function init_commands(client)
+  local trigger_completion_cmd = 'editor.action.triggerSuggest'
+
+  -- Check if the command is in the global registry or in the client's commands.
+  if not lsp.commands[trigger_completion_cmd] and not client.commands[trigger_completion_cmd] then
+    client.commands[trigger_completion_cmd] = function()
+      pcall(M.trigger)
+    end
+  end
+end
```

Comment:
```suggestion
```

This is off-spec, we can try adding this later, together with a PR against the lsp repo to add it to the spec.